### PR TITLE
Ocaml lsp improvements

### PR DESCRIPTION
**Pull request by rgrinberg**
###### Motivation for this change

specify the dependencies of ocamllsp and jsonrpc more accurately

###### Things done

* Dropped unnecessary dependencies for jsonrpc 1.7.0
* Fixed the minimum OCaml version for ocamllsp 1.7.0
* Fixed the set of dependencies ofr ocamllsp 1.7.0

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rgrinberg** [comment]
cc @vbgl

**Comment by vbgl** [comment]
I don’t understand the second commit.

Please double check [the contribution guidelines](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md) and include the attribute path of the modified package on the first line of the commit message.

**Comment by rgrinberg** [comment]
> I don’t understand the second commit.

ocaml-lsp 1.7.0 does not support 4.06 as was erroneously stated before. The new version only supports 4.12, as the [opam file](https://github.com/ocaml/ocaml-lsp/blob/master/ocaml-lsp-server.opam#L33) confirms.

> Please double check the contribution guidelines and include the attribute path of the modified package on the first line of the commit message.

Will do.

**Comment by rgrinberg** [comment]
@vbgl updated the commits to follow the convention.

**Comment by vbgl** [comment]
Thanks. The second commit still looks like obfuscation to me, but I did probably not get its purpose.

File: pkgs/development/ocaml-modules/ocaml-lsp/jsonrpc.nix [hunk 1]
```diff
@@ -31,9 +31,17 @@ buildDunePackage rec {
   useDune2 = true;
   minimumOCamlVersion = "4.06";
 
-  buildInputs = [ yojson stdlib-shims ocaml-syntax-shims ];
+  buildInputs =
+    if lib.versionAtLeast version "1.7.0" then
+      [ ]
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/ocaml-lsp/jsonrpc.nix:36]: Why not use lib.optional?

Comment (rgrinberg) [reply pkgs/development/ocaml-modules/ocaml-lsp/jsonrpc.nix:36]: In my opinion, it's harder for a human to read the list of dependency when one also has to evaluate inline expressions.

**Pull request merged**